pkg/k8splan/watcher.go

```diff
@@ -0,0 +1,198 @@
+package k8splan
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/rancher/lasso/pkg/cache"
+	"github.com/rancher/lasso/pkg/client"
+	"github.com/rancher/lasso/pkg/controller"
+	"github.com/rancher/system-agent/pkg/applyinator"
+	"github.com/rancher/system-agent/pkg/config"
+	"github.com/rancher/system-agent/pkg/prober"
+	corecontrollers "github.com/rancher/wrangler/pkg/generated/controllers/core/v1"
+	"github.com/sirupsen/logrus"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/tools/clientcmd"
+)
+
+const (
+	appliedChecksumKey   = "applied-checksum"
+	appliedOutputKey     = "applied-output"
+	probeStatusesKey     = "probe-statuses"
+	probePeriodKey       = "probe-period"
+	planKey              = "plan"
+	enqueueAfterDuration = "5s"
+)
+
+func Watch(ctx context.Context, applyinator applyinator.Applyinator, connInfo config.ConnectionInfo) error {
+	w := &watcher{
+		connInfo:    connInfo,
+		applyinator: applyinator,
+	}
+
+	go w.start(ctx)
+
+	return nil
+}
+
+type watcher struct {
+	connInfo                   config.ConnectionInfo
+	applyinator                applyinator.Applyinator
+	lastAppliedResourceVersion string
+}
+
+func (w *watcher) start(ctx context.Context) {
+	kc, err := clientcmd.RESTConfigFromKubeConfig([]byte(w.connInfo.KubeConfig))
+	if err != nil {
+		panic(err)
+	}
+
+	clientFactory, err := client.NewSharedClientFactory(kc, nil)
+	if err != nil {
+		panic(err)
+	}
+
+	cacheFactory := cache.NewSharedCachedFactory(clientFactory, &cache.SharedCacheFactoryOptions{
+		DefaultNamespace: w.connInfo.Namespace,
+		DefaultTweakList: func(options *metav1.ListOptions) {
+			options.FieldSelector = fmt.Sprintf("metadata.name=%s", w.connInfo.SecretName)
+		},
+	})
+
+	controllerFactory := controller.NewSharedControllerFactory(cacheFactory, nil)
+	core := corecontrollers.New(controllerFactory)
+
+	probePeriod, err := time.ParseDuration(enqueueAfterDuration)
+	if err != nil {
+		panic(err)
+	}
+
+	core.Secret().OnChange(ctx, "secret-watch", func(s string, secret *v1.Secret) (*v1.Secret, error) {
+		if rawPeriod, ok := secret.Data[probePeriodKey]; ok {
+			if parsedPeriod, err := time.ParseDuration(fmt.Sprintf("%ss", string(rawPeriod))); err != nil {
+				logrus.Errorf("[K8s] error parsing duration %ss, using default", string(rawPeriod))
+			} else {
+				probePeriod = parsedPeriod
+			}
+		}
+		if secret == nil {
+			logrus.Debugf("[K8s] Secret was nil")
+			core.Secret().EnqueueAfter(w.connInfo.Namespace, w.connInfo.SecretName, probePeriod)
+			return secret, nil
+		}
+		secret = secret.DeepCopy()
+		logrus.Debugf("[K8s] Processing secret %s in namespace %s at generation %d with resource version %s", secret.Name, secret.Namespace, secret.Generation, secret.ResourceVersion)
```

Comment:
Would it be clearer if we changed `[K8s]` to `[system-agent]`?

Reply:
The reason for `[K8s]` is to help distinguish between `[K8s]` and `[local]` (source of plan data)